File: 3575.maximum-good-subtree-score.py
#
# @lc app=leetcode id=3575 lang=python3
#
# [3575] Maximum Good Subtree Score
#

from typing import List

# @lc code=start
class Solution:
    def goodSubtreeSum(self, vals: List[int], par: List[int]) -> int:
        n = len(vals)
        adj = [[] for _ in range(n)]
        for i in range(1, n):
            adj[par[i]].append(i)
        MOD = 10**9 + 7
        INF = 10**18 + 10**12
        total = [0]

        def get_mask(val: int) -> int:
            seen = 0
            s = str(val)
            for ch in s:
                d = ord(ch) - ord('0')
                if seen & (1 << d):
                    return -1
                seen |= (1 << d)
            return seen

        def dfs(u: int) -> List[int]:
            mask_u = get_mask(vals[u])
            dp_children = [-INF] * 1024
            dp_children[0] = 0
            for v in adj[u]:
                child_dp = dfs(v)
                new_dp = [-INF] * 1024
                active_prev = [p for p in range(1024) if dp_children[p] != -INF]
                active_child = [c for c in range(1024) if child_dp[c] != -INF]
                for p in active_prev:
                    ps = dp_children[p]
                    for c in active_child:
                        if (p & c) == 0:
                            nm = p | c
                            new_dp[nm] = max(new_dp[nm], ps + child_dp[c])
                dp_children = new_dp
            # dp_u without u
            dp_u = dp_children[:]
            # add u if possible
            if mask_u != -1:
                val_u = vals[u]
                active_here = [m for m in range(1024) if dp_children[m] != -INF]
                for m in active_here:
                    if (m & mask_u) == 0:
                        nm = m | mask_u
                        dp_u[nm] = max(dp_u[nm], dp_children[m] + val_u)
            # max_sc
            max_sc = 0
            for i in range(1024):
                if dp_u[i] > max_sc:
                    max_sc = dp_u[i]
            total[0] = (total[0] + max_sc) % MOD
            return dp_u

        dfs(0)
        return total[0]

# @lc code=end

File: 3575.maximum-good-subtree-score.cpp
#include <vector>
#include <algorithm>
#include <climits>
using namespace std;

class Solution {
public:
    static constexpr long long MOD = 1000000007LL;
    long long answer;

    vector<long long> dfs(int u,
                         const vector<int>& vals,
                         const vector<vector<int>>& adj) {
        const int FULL_MASK = (1 << 10) - 1;
        vector<long long> dp(1024, LLONG_MIN);
        dp[0] = 0;

        bool valid_node = true;
        int num = vals[u];
        bool seen[10] = {false};
        int mask_u = 0;
        while (num > 0) {
            int d = num % 10;
            if (seen[d]) {
                valid_node = false;
                break;
            }
            seen[d] = true;
            mask_u |= (1 << d);
            num /= 10;
        }

        if (valid_node)
            dp[mask_u] = max(dp[mask_u], static_cast<long long>(vals[u]));

        for (int v : adj[u]) {
            auto child_dp = dfs(v, vals, adj);
            vector<long long> new_dp(1024, LLONG_MIN);

            vector<int> active_parent;
            for (int m = 0; m < 1024; ++m)
                if (dp[m] != LLONG_MIN)
                    active_parent.push_back(m);

            const int FULL = FULL_MASK;
            for (int m_parent : active_parent) {
                long long sparent = dp[m_parent];
                int complement = (~m_parent) & FULL;
                int submask = complement;
                while (true) {
                    long long schild = child_dp[submask];
                    if (schild != LLONG_MIN) {
                        long long total = sparent + schild;
                        int new_mask = m_parent | submask;
                        if (total > new_dp[new_mask])
                            new_dp[new_mask] = total;
                    }
                    if (submask == 0)
                        break;
                    submask = (submask - 1) & complement;
                }
            }
            dp.swap(new_dp);
        }

        long long best_sum_subtree = *max_element(dp.begin(), dp.end());
        answer += best_sum_subtree % MOD;
        answer %= MOD;

        return dp;
    }

    int goodSubtreeSum(vector<int>& vals_vec,
                       vector<int>& par_vec) {
        answer = 0LL;
        int n = vals_vec.size();
        vector<vector<int>> adj(n);
        
        for(int i = 0;i < n;i++) {
            int p = par_vec[i];
            if(p != -1)
                adj[p].push_back(i);
        }

        dfs(0,
             vals_vec,
             adj);

        return answer % MOD;
    }
};